firebase-firestore/src/androidTest/java/com/google/firebase/firestore/AggregationTest.java

```diff
@@ -0,0 +1,1030 @@
+          "x${collectionGroup}/not-cg-doc",
+          "${collectionGroup}x/not-cg-doc",
+          "abc/123/${collectionGroup}x/not-cg-doc",
+          "abc/123/x${collectionGroup}/not-cg-doc",
+          "abc/${collectionGroup}"
+        };
+    WriteBatch batch = db.batch();
+    for (String path : docPaths) {
+      batch.set(db.document(path.replace("${collectionGroup}", collectionGroup)), map("x", 2));
+    }
+    waitFor(batch.commit());
+
+    AggregateQuerySnapshot snapshot =
+        waitFor(
+            db.collectionGroup(collectionGroup)
+                .aggregate(AggregateField.count(), sum("x"), average("x"))
+                .get(AggregateSource.SERVER));
+    assertEquals(
+        5L, // "cg-doc1", "cg-doc2", "cg-doc3", "cg-doc4", "cg-doc5",
+        snapshot.get(AggregateField.count()));
+    assertEquals(10L, snapshot.get(sum("x")));
+    assertEquals((Double) 2.0, snapshot.get(average("x")));
+  }
+
+  @Test
+  public void testPerformsAggregationsWhenNaNExistsForSomeFieldValues() {
+    Map<String, Map<String, Object>> testDocs =
+        map(
+            "a",
+            map("author", "authorA", "title", "titleA", "pages", 100, "year", 1980, "rating", 5),
+            "b",
+            map("author", "authorB", "title", "titleB", "pages", 50, "year", 2020, "rating", 4),
+            "c",
+            map(
+                "author",
+                "authorC",
+                "title",
+                "titleC",
+                "pages",
+                100,
+                "year",
+                1980,
+                "rating",
+                Double.NaN),
+            "d",
+            map("author", "authorD", "title", "titleD", "pages", 50, "year", 2020, "rating", 0));
+    CollectionReference collection = testCollectionWithDocs(testDocs);
+
+    AggregateQuerySnapshot snapshot =
+        waitFor(
+            collection
+                .aggregate(sum("rating"), sum("pages"), average("year"), average("rating"))
+                .get(AggregateSource.SERVER));
+
+    assertEquals(snapshot.get(sum("rating")), Double.NaN);
+    assertEquals(snapshot.get(sum("pages")), 300L);
+    assertEquals(snapshot.get(average("rating")), (Double) Double.NaN);
+    assertEquals(snapshot.get(average("year")), (Double) 2000.0);
+  }
+
+  @Test
+  public void testThrowsAnErrorWhenGettingTheResultOfAnUnrequestedAggregation() {
+    CollectionReference collection = testCollectionWithDocs(testDocs1);
+
+    AggregateQuerySnapshot snapshot =
+        waitFor(collection.aggregate(sum("pages")).get(AggregateSource.SERVER));
+
+    Exception exception = null;
+    try {
+      snapshot.get(average("pages"));
+    } catch (Exception e) {
+      exception = e;
+    }
+    assertNotNull(exception);
+    assertEquals(
+        exception.getMessage(), "'average(pages)' was not requested in the aggregation query.");
+
+    exception = null;
+    try {
+      snapshot.get(sum("foo"));
+    } catch (Exception e) {
```

Comment:
We want to assert that the thrown exception type is `RuntimeException` per the API spec